numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -162,6 +164,46 @@ PyArray_OptionalIntpConverter(PyObject *obj, PyArray_Dims *seq)
     return PyArray_IntpConverter(obj, seq);
 }
 
+NPY_NO_EXPORT int
+PyArray_CopyConverter(PyObject *obj, PyNpCopyMode_Enum *copymode) {
+    if (obj == Py_None) {
+        PyErr_SetString(PyExc_ValueError,
+                        "NoneType copy mode not allowed. Please choose one of " 
+                         "np.CopyMode.ALWAYS, np.CopyMode.IF_NEEDED, np.CopyMode.NEVER.");
+        return NPY_FAIL;
+    }
+
+    int int_copymode = -1;
+    PyObject* numpy_CopyMode = NULL;
+    npy_cache_import("numpy._globals", "CopyMode", &numpy_CopyMode);
+    if( numpy_CopyMode != NULL ) {
+        if(PyObject_IsInstance(obj, numpy_CopyMode)) {
+            PyObject* mode_value = PyObject_GetAttrString(obj, "value");
+            PyArray_PythonPyIntFromInt(mode_value, &int_copymode);
+        }
+    }
```

Comment:
```suggestion
    if (numpy_CopyMode == NULL) {
        return NPY_FAIL;
    }
    if (PyObject_IsInstance(obj, numpy_CopyMode)) {
        PyObject* mode_value = PyObject_GetAttrString(obj, "value");
        PyArray_PythonPyIntFromInt(mode_value, &int_copymode);
    }
```

Reply:
Are there reasons against treating them as singletons?
```
if value is np.CopyMode.NEVER:
    ...
elif value is np.CopyMode.IF_NEEDED:
   ...
elif value is np.CopyMode.FORCE:
   ...
elif None:
   return  # (not sure we allow None for default)
```

Reply:
So, what happens in the above code is, first we try to verify if the copy argument is an instance of `np.CopyMode`. If it is then its value extracted and stored in `int_copymode`. If that doesn't work, then we follow the conventional path i.e., extract value of `obj` node via, `PyArray_PythonPyIntFromInt`. This helps in two ways, i) User can pass, `np.CopyMode.NEVER` and ii) They can also pass the values currently supported in `main` branch like, `True`, `False`, etc. In other words, the new changes extend the API and don't break anything existing.

Now, in the `if-elif` ladder in your comment, may break for values of copy argument which are currently supported since, we are just checking whether value is same as one of the options under `np.CopyMode`.

Reply:
Yeah, the last should just have been `else: current_behaviour`. But I think in Python we would write `is` also?  And in that case it seems nicer to me here as well (you can add the isinstance/type-check check, but if its a 3-way if that should not be worthwhile).

Reply:
One another way to simplify the checks would be to first import, `np.CopyMode.ALWAYS`, `np.CopyMode.IF_NEEDED`, `np.CopyMode.NEVER`, `np.True_`, `np.False_` and then check whether the object is equal to any of these values, `True` or `False` using, `PyObject_RichCompareBool`. What do you say?